core/services/gas/models.go

```diff
@@ -0,0 +1,213 @@
+// Transaction represents an ethereum transaction
+// Use our own type because geth's type has validation failures on e.g. zero
+// gas used, which can occur on other chains.
+// This type is only used for the gas updater, and can be expensive to unmarshal. Don't add unnecessary fields here.
+type Transaction struct {
+	GasPrice *big.Int
+	GasLimit uint64
+}
+
+// UnmarshalJSON unmarshals a Transaction
+func (t *Transaction) UnmarshalJSON(data []byte) error {
+	ti := transactionInternal{}
+	if err := json.Unmarshal(data, &ti); err != nil {
+		return errors.Wrapf(err, "failed to unmarshal to transactionInternal, got: '%s'", data)
+	}
+	if ti.Gas == nil {
+		return errors.Errorf("expected 'gas' to not be null, got: '%s'", data)
+	}
+	*t = Transaction{
+		(*big.Int)(ti.GasPrice),
+		uint64(*ti.Gas),
+	}
+	return nil
+}
+
+func bumpGasPriceOnly(config Config, originalGasPrice *big.Int, originalGasLimit uint64) (gasPrice *big.Int, chainSpecificGasLimit uint64, err error) {
+	gasPrice, err = bumpGasPrice(config, originalGasPrice)
+	if err != nil {
+		return nil, 0, err
+	}
+	chainSpecificGasLimit = originalGasLimit
+	return
+}
+
+// TODO: Move this?
+// bumpGasPrice computes the next gas price to attempt as the largest of:
+// - A configured percentage bump (ETH_GAS_BUMP_PERCENT) on top of the baseline price.
+// - A configured fixed amount of Wei (ETH_GAS_PRICE_WEI) on top of the baseline price.
+// The baseline price is the maximum of the previous gas price attempt and the node's current gas price.
+func bumpGasPrice(config Config, originalGasPrice *big.Int) (*big.Int, error) {
+	baselinePrice := max(originalGasPrice, config.EthGasPriceDefault())
+
+	var priceByPercentage = new(big.Int)
+	priceByPercentage.Mul(baselinePrice, big.NewInt(int64(100+config.EthGasBumpPercent())))
+	priceByPercentage.Div(priceByPercentage, big.NewInt(100))
+
+	var priceByIncrement = new(big.Int)
+	priceByIncrement.Add(baselinePrice, config.EthGasBumpWei())
+
+	bumpedGasPrice := max(priceByPercentage, priceByIncrement)
+	if bumpedGasPrice.Cmp(config.EthMaxGasPriceWei()) > 0 {
+		// promGasBumpExceedsLimit.Inc()
+		return config.EthMaxGasPriceWei(), errors.Errorf("bumped gas price of %s would exceed configured max gas price of %s (original price was %s). %s",
+			bumpedGasPrice.String(), config.EthMaxGasPriceWei(), originalGasPrice.String(), static.EthNodeConnectivityProblemLabel)
+	} else if bumpedGasPrice.Cmp(originalGasPrice) == 0 {
+		// NOTE: This really shouldn't happen since we enforce minimums for
+		// ETH_GAS_BUMP_PERCENT and ETH_GAS_BUMP_WEI in the config validation,
+		// but it's here anyway for a "belts and braces" approach
+		return bumpedGasPrice, errors.Errorf("bumped gas price of %s is equal to original gas price of %s."+
+			" ACTION REQUIRED: This is a configuration error, you must increase either "+
+			"ETH_GAS_BUMP_PERCENT or ETH_GAS_BUMP_WEI", bumpedGasPrice.String(), originalGasPrice.String())
+	}
+	// TODO: Move/fix these
+	// promNumGasBumps.Inc()
```

Comment:
could be moved to the block history estimator?

Reply:
Decided to keep it here to maintain compatibility